src/ext/serializer.c

```diff
@@ -173,32 +175,125 @@ int ddtrace_serialize_simple_array(zval *trace, zval *retval TSRMLS_DC) {
+    zend_call_method_with_0_params(&exception, Z_OBJCE_P(exception), NULL, "gettraceasstring", &stack);
+
+    /* add_assoc_stringl does not actually mutate the string, but we've either
+     * already made a copy, or it will when it duplicates with dup param, so
+     * if it did it should still be safe. */
+    add_assoc_stringl(meta, "error.name", (char *)class_name, class_name_len, dup);
+    add_assoc_zval(meta, "error.msg", msg);
+    add_assoc_zval(meta, "error.stack", stack);
+}
+
+void _serialize_meta(zval *el, ddtrace_span_t *span TSRMLS_DC) {
+    zval *meta = _read_span_property(span->span_data, "meta", sizeof("meta") - 1 TSRMLS_CC);
+    BOOL_T should_free = TRUE;
+
+    if (!meta || Z_TYPE_P(meta) != IS_ARRAY) {
+        ALLOC_INIT_ZVAL(meta);
+        array_init(meta);
+    } else {
+        should_free = FALSE;
+        Z_ADDREF_P(meta);
+    }
+
+    if (span->exception) {
+        _serialize_exception(el, meta, span TSRMLS_CC);
+    }
+
+    // Add meta only if it has elements
+    if (zend_hash_num_elements(Z_ARRVAL_P(meta))) {
+        add_assoc_zval(el, "meta", meta);
+    } else {
+        zval_dtor(meta);
+
+        if (should_free) {
+            efree(meta);
+        }
+    }
+}
+
 #else
-#define ADD_ELEMENT_IF_PROP_TYPE(name, type)                                                                         \
-    do {                                                                                                             \
-        zval rv;                                                                                                     \
-        zval *prop =                                                                                                 \
-            zend_read_property(ddtrace_ce_span_data, span->span_data, (name), sizeof((name)) - 1, 1, &rv TSRMLS_CC); \
-        if (Z_TYPE_P(prop) == (type)) {                                                                              \
-            zval value;                                                                                              \
-            ZVAL_COPY(&value, prop);                                                                                 \
-            add_assoc_zval(el, (name), &value);                                                                      \
-        }                                                                                                            \
-    } while (0);
+static zval *_read_span_property(zval *span_data, const char *name, size_t name_len) {
+    zval rv;
+    return zend_read_property(ddtrace_ce_span_data, span_data, name, name_len, 1, &rv);
+}
+
+static void _add_assoc_zval(zval *el, const char *name, zval *prop) {
+    zval value;
+    ZVAL_COPY(&value, prop);
+    add_assoc_zval(el, (name), &value);
+}
+
+void _serialize_exception(zval *el, zval *meta, ddtrace_span_t *span) {
+    zval exception, name, msg, stack;
+    ZVAL_OBJ(&exception, span->exception);
```

Comment:
Would it make sense to check if `span->exception` is`null` here as in both `span->exception == NULL && Z_TYPE_P(tmp) == IS_NULL)`?